File: scrape/price.py
"""
Update the database with the latest prices.

CRON JOB: 0 06 1 * *
"""

import os
import enum
from typing import List
import concurrent.futures
from datetime import datetime, timedelta

import pymongo
import requests
from pymongo.errors import BulkWriteError
from pymongo.results import BulkWriteResult
from pymongo.mongo_client import MongoClient


_CLIENT = MongoClient(
    f'mongodb+srv://{os.environ.get("MONGO_USR")}:{os.environ.get("MONGO_PWD")}'
    f'@vinskraper.wykjrgz.mongodb.net/'
    f'?retryWrites=true&w=majority&appName=vinskraper'
)
_DATABASE = _CLIENT['vinskraper']['varer']

_URL = ('https://www.vinmonopolet.no/vmpws/v2/vmp/'
        'search?searchType=product'
        '&currentPage={}'
        '&q=%3Arelevance%3AmainCategory%3A{}')

_PROXIES = iter([
    {
        "http": f"http://{os.environ.get('PROXY_USR')}:{os.environ.get('PROXY_PWD')}@{ip}:{os.environ.get('PROXY_PRT')}"
    }
    for ip in os.environ.get('PROXY_IPS', '').split(',')
])
_PROXY = next(_PROXIES)

_SESSION = requests.Session()

_OLD = (datetime.now() - timedelta(days=15)).strftime("%Y-%m-01")
_NOW = datetime.now().strftime('%Y-%m-01')
_IMAGE = {'thumbnail': 'https://bilder.vinmonopolet.no/bottle.png',
          'product': 'https://bilder.vinmonopolet.no/bottle.png'}


class CATEGORY(enum.Enum):
    """
    Enum class for the different categories of products available at vinmonopolet.
    Extends the `scrape._URL` with the category value.
    """
    RED_WINE = 'rødvin'
    WHITE_WINE = 'hvitvin'
    SPARKLING_WINE = 'musserende_vin'
    PEARLING_WINE = 'perlende_vin'
    FORTIFIED_WINE = 'sterkvin'
    AROMATIC_WINE = 'aromatisert_vin'
    FRUIT_WINE = 'fruktvin'
    ROSE_WINE = 'rosévin'
    SPIRIT = 'brennevin'
    BEER = 'øl'
    CIDER = 'sider'
    SAKE = 'sake'
    MEAD = 'mjød'
    ALCOHOL_FREE = 'alkoholfritt'


def _upsert(data: List[dict]) -> BulkWriteResult:
    """
    Upsert the given data into the database.
    Calculates the price change percentage based on the previous and current price.

    Parameters
    ----------
    data : List[dict]
        The data to insert into the database.

    Returns
    -------
    BulkWriteResult
    """
    operations = [
        pymongo.UpdateOne(
            {'index': record['index']},
            [
                {'$set': record},
                {'$set': {
                    f'prisendring {_NOW}': {'$cond': [
                        {'$gt': [f'$pris {_OLD}', 0]},
                        {'$multiply': [
                            {'$divide': [
                                {'$subtract': [record[f'pris {_NOW}'], f'$pris {_OLD}']},
                                f'$pris {_OLD}'
                            ]},
                            100
                        ]},
                        0
                    ]},
                }},
                {'$set': {
                    'literpris': {'$cond': {
                        'if': {'$and': [
                            {'$ne': [f'$pris {_NOW}', None]},
                            {'$ne': [f'$pris {_NOW}', 0]},
                            {'$ne': ['$volum', None]},
                            {'$ne': ['$volum', 0]},
                        ]},
                        'then': {'$multiply': [
                            {'$divide': [f'$pris {_NOW}', '$volum']},
                            100
                        ]},
                        'else': None
                    }}
                }},
                {'$set': {
                    'alkoholpris': {'$cond': {''
                        'if': {'$and': [
                            {'$ne': ['$literpris', None]},
                            {'$ne': ['$literpris', 0]},
                            {'$ne': ['$alkohol', None]},
                            {'$ne': ['$alkohol', 0]},
                        ]},
                        'then': {'$divide': ['$literpris', '$alkohol']},
                        'else': None
                    }}
                }}
            ],
            upsert=True
        )
        for record in data
    ]
    result = _DATABASE.bulk_write(operations)

    return result


def _scrape_page(
        category: CATEGORY,
        page: int
) -> requests.models.Response:
    """
    Scrape a single page of products from the given category.

    Parameters
    ----------
    category : CATEGORY
        The category of products to fetch.
    page : int
        The page number to fetch.

    Returns
    -------
    requests.models.Response
        The response from the page.

    Notes
    -----
        Tries to fetch the page using the `_PROXY`.
        If it fails, it renews the proxy (`_PROXY.renew()`) and retries.
        Returns code `500` if it fails 10 consecutive times.
    """
    global _PROXY

    for _ in range(10):
        try:
            response = _SESSION.get(_URL.format(page, category.value),
                                    proxies=_PROXY,
                                    timeout=3)
            break
        except Exception as err:
            print(f'Error: Page {page} (trying another proxy); {err}')
            try:
                _PROXY = next(_PROXIES)
            except StopIteration:
                raise ValueError('Failed to fetch new products. No more proxies.')
    else:
        print(f'Error: Failed to fetch page {page} after 10 attempts.')
        response = requests.models.Response()
        response.status_code = 500

    return response


def _process_images(images):
    return {img['format']: img['url'] for img in images} if images else _IMAGE


def _process(products) -> List[dict]:
    return [{
        'index': int(product.get('code', 0)),

        'status': product.get('status', None),
        'kan kjøpes': product.get('buyable', False),
        'utgått': product.get('expired', True),

        'tilgjengelig for bestilling': product.get(
            'productAvailability', {}
        ).get(
            'deliveryAvailability', {}
        ).get(
            'availableForPurchase', False
        ),

        'bestillingsinformasjon': product.get(
            'productAvailability', {}
        ).get(
            'deliveryAvailability', {}
        ).get(
            'infos', [{}]
        )[0].get(
            'readableValue', None
        ),

        'tilgjengelig i butikk': product.get(
            'productAvailability', {}
        ).get(
            'storesAvailability', {}
        ).get(
            'availableForPurchase', False
        ),

        'butikkinformasjon': product.get(
            'productAvailability', {}
        ).get(
            'storesAvailability', {}
        ).get('infos', [{}]
        )[0].get(
            'readableValue', None
        ),

        f'pris {_NOW}': product.get('price', {}).get('value', 0.0),
    } for product in products]


def _scrape_category(category: CATEGORY) -> bool:
    """
    Fetches all products from the given category.

    Parameters
    ----------
    category : CATEGORY
        The category of products to fetch.

    Returns
    -------
    bool

    Notes
    -----
        Loops through the pages of the category.
        Assumes that there is less than 10000 pages.
        Parses the response and extracts the products.
        Extracts the product information and stores it in the dictionary.
        The price key is suffixed with the current (year-month-)timestamp (%Y-%m-01).
    """

    # ----------------------------------------------------------------------------------------------
    # Loops through the pages of the category.
    # Assuming that there is less than 10000 pages.

    items = []
    for page in range(10000):
        products = _scrape_page(category, page)

        if products.status_code != 200:
            print(f'{category.name} Failed: '
                  f'Page {page} (status code {products.status_code}): '
                  f'{products.text}.')
            continue

        # ------------------------------------------------------------------------------------------
        # Parses the response and extracts the products.

        products = products.json().get('productSearchResult', {}).get('products', [])
        if not products:
            print(f'{category.name} No more products (final page: {page - 1}).')
            break

        # ------------------------------------------------------------------------------------------
        # Extracts the product information and stores it in the list.

        items.extend(_process(products))

    print(f'{category.name} Inserting into database.')
    try:
        result = _upsert(items)
    except BulkWriteError as bwe:
        print(f'{category.name} Error: Bulk write operation; {bwe.details}.')
        return False
    print(f'{category.name} Modified {result.modified_count} records')
    print(f'{category.name} Upserted {result.upserted_count} records')
    print(f'{category.name} Success.')
    return True


def scrape(categories=None, max_workers=10):
    if categories is None:
        categories = list(CATEGORY)
    failed = []

    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
        results = {executor.submit(_scrape_category, category): category
                   for category in categories}

        for future in concurrent.futures.as_completed(results):
            try:
                success = future.result()
                if not success:
                    failed.append(results[future])
            except Exception as exc:
                category = results[future]
                print(f'{category.name} FAILED! Concurrency error: {exc}')
                failed.append(category)

    expired = list(_DATABASE.find({'utgått': True}))
    if expired:
        _CLIENT['vinskraper']['utgått'].insert_many(expired)

        ids = [doc['index'] for doc in expired]
        result = _DATABASE.delete_many({'index': {'$in': ids}})

        print(f"Moved {len(ids)} documents to the expired collection.")
        print(f"Deleted {result.deleted_count} documents from the stock collection.")
    else:
        print("No expired documents found.")

    return failed

File: scrape/price.mjs
import axios from "axios";
import { HttpsProxyAgent } from "https-proxy-agent";
import { MongoClient, ServerApiVersion } from "mongodb";
import dotenv from "dotenv";

dotenv.config();

const client = new MongoClient(
  `mongodb+srv://${process.env.MONGO_USR}:${process.env.MONGO_PWD}@snublejuice.faktu.mongodb.net/?retryWrites=true&w=majority&appName=snublejuice`,
  {
    serverApi: {
      version: ServerApiVersion.v1,
      strict: true,
      deprecationErrors: true,
    },
  },
);
await client.connect();

const database = client.db("snublejuice");
const itemCollection = database.collection("products");

const proxies = process.env.PROXY_IPS.split(",").flatMap((ip) => [
  new HttpsProxyAgent(
    `http://${process.env.PROXY_USR}:${process.env.PROXY_PWD}@${ip}:${process.env.PROXY_PRT}`,
  ),
]);

function getNextProxy(proxy) {
  if (proxies.length === 0) {
    throw new Error("No more proxies available!");
  }

  const index = proxy ? proxies.findIndex((p) => p.proxy.host === proxy.proxy.host) : -1;
  return proxies[(index + 1) % proxies.length];
}

function removeProxy(proxy) {
  const index = proxies.findIndex((p) => p.proxy.host === proxy.proxy.host);
  if (index > -1) {
    const removedProxy = proxies.splice(index, 1)[0];
    console.log(
      `Removed failing proxy ${removedProxy.proxy.host}. ${proxies.length} proxies remaining.`,
    );
  }
}

const URL =
  "https://www.vinmonopolet.no/vmpws/v2/vmp/search?fields=FULL&searchType=product&currentPage={}&q=%3Arelevance";
const LINK = "https://www.vinmonopolet.no{}";

const IMAGE = {
  thumbnail: "https://bilder.vinmonopolet.no/bottle.png",
  product: "https://bilder.vinmonopolet.no/bottle.png",
};

function processImages(images) {
  return images ? images.reduce((acc, img) => ({ ...acc, [img.format]: img.url }), {}) : IMAGE;
}

function processProducts(products, alreadyUpdated) {
  const processed = [];

  for (const product of products) {
    const index = parseInt(product.code, 10) || null;

    // Extra check to avoid duplicates.
    // The alreadyUpdated is set in the main function below.
    // Only used if the scraping crashes and needs to be restarted.
    // In this way, it skips the already processed products (before crash).
    if (alreadyUpdated.includes(index)) {
      continue;
    }

    processed.push({
      index: index,

      updated: true,

      name: product.name || null,
      price: product.price?.value || 0.0,

      volume: product.volume?.value || 0.0,
      literprice:
        product.price?.value && product.volume?.value
          ? product.price.value / (product.volume.value / 100.0)
          : 0.0,

      url: product.url ? LINK.replace("{}", product.url) : null,
      images: product.images ? processImages(product.images) : IMAGE,

      category: product.main_category?.name || null,
      subcategory: product.main_sub_category?.name || null,

      country: product.main_country?.name || null,
      district: product.district?.name || null,
      subdistrict: product.sub_District?.name || null,

      selection: product.product_selection || null,
      sustainable: product.sustainable || false,

      buyable: product.buyable || false,
      expired: product.expired || true,
      status: product.status || null,

      orderable: product.productAvailability?.deliveryAvailability?.availableForPurchase || false,
      orderinfo:
        product.productAvailability?.deliveryAvailability?.infos?.[0]?.readableValue || null,

      instores: product.productAvailability?.storesAvailability?.availableForPurchase || false,
      storeinfo: product.productAvailability?.storesAvailability?.infos?.[0]?.readableValue || null,
    });
  }

  return processed;
}

async function getPage(page, proxy, alreadyUpdated) {
  for (let attempt = 0; attempt < 5; attempt++) {
    try {
      const response = await session.get(URL.replace("{}", page), {
        httpsAgent: proxy,
        timeout: 10000,
      });

      if (response.status === 200) {
        return processProducts(response.data["productSearchResult"]["products"], alreadyUpdated);
      }

      console.log(`Status code ${response.status} at page ${page} (trying another proxy); ${err}`);
    } catch (err) {
      console.log(`Request failed with proxy ${proxy.host} for page ${page}: ${err.message}`);
      if (err.code === "ECONNREFUSED" || err.code === "ETIMEDOUT") {
        removeProxy(proxy);
      }
    }

    try {
      proxy = getNextProxy(proxy);
      await new Promise((resolve) => setTimeout(resolve, 2000));
    } catch (err) {
      throw new Error(`No more proxies available to fetch id ${id}`);
    }
  }

  throw new Error(`Failed to fetch page ${page} after 5 attempts.`);
}

async function updateDatabase(data) {
  const operations = data.map((record) => ({
    updateOne: {
      filter: { index: record.index },
      update: [
        { $set: { oldprice: "$price" } },
        { $set: record },
        { $set: { prices: { $ifNull: ["$prices", []] } } },
        { $set: { prices: { $concatArrays: ["$prices", ["$price"]] } } },
        {
          $set: {
            discount: {
              $cond: {
                if: {
                  $and: [
                    { $gt: ["$oldprice", 0] },
                    { $gt: ["$price", 0] },
                    { $ne: ["$oldprice", null] },
                    { $ne: ["$price", null] },
                  ],
                },
                then: {
                  $multiply: [
                    {
                      $divide: [{ $subtract: ["$price", "$oldprice"] }, "$oldprice"],
                    },
                    100,
                  ],
                },
                else: 0,
              },
            },
            literprice: {
              $cond: {
                if: {
                  $and: [
                    { $gt: ["$price", 0] },
                    { $gt: ["$volume", 0] },
                    { $ne: ["$price", null] },
                    { $ne: ["$volume", null] },
                  ],
                },
                then: {
                  $multiply: [
                    {
                      $divide: ["$price", "$volume"],
                    },
                    100,
                  ],
                },
                else: null,
              },
            },
          },
        },
        {
          $set: {
            alcoholprice: {
              $cond: {
                if: {
                  $and: [
                    { $gt: ["$literprice", 0] },
                    { $gt: ["$alcohol", 0] },
                    { $ne: ["$literprice", null] },
                    { $ne: ["$alcohol", null] },
                  ],
                },
                then: {
                  $divide: ["$literprice", "$alcohol"],
                },
                else: null,
              },
            },
          },
        },
      ],
      upsert: true,
    },
  }));

  return await itemCollection.bulkWrite(operations);
}

async function getProducts(startPage = 0, alreadyUpdated = []) {
  let items = [];
  let proxy = getNextProxy();

  for (let page = startPage; page < 10000; page++) {
    // Fetch products of page.
    console.log(`Page ${page}.`);
    try {
      let products = await getPage(page, proxy, alreadyUpdated);
      if (products.length === 0) {
        console.log(`No more products (final page: ${page - 1}).`);
        break;
      }

      items = items.concat(products);
      proxy = getNextProxy();

      await new Promise((resolve) => setTimeout(resolve, 900));
    } catch (err) {
      console.log(`Error page ${page}! ${err}`);
      break;
    }

    // Upsert to the database every 10 pages.
    if (page % 10 === 0) {
      if (items.length === 0) {
        throw new Error(`No items for the last 10 pages. Aborting.`);
      }

      console.log(`Updating ${items.length} products.`);
      const result = await updateDatabase(items);
      console.log(` Modified ${result.modifiedCount} records`);
      console.log(` Upserted ${result.upsertedCount} records`);

      items = [];
    }
  }

  // Upsert the remaining products, if any.
  if (items.length === 0) {
    return;
  }
  console.log(`Updating ${items.length} final products.`);
  const result = await updateDatabase(items);
  console.log(` Modified ${result.modifiedCount} records`);
  console.log(` Upserted ${result.upsertedCount} records`);
}

async function syncUnupdatedProducts(threshold = null) {
  const unupdatedCount = await itemCollection.countDocuments({ updated: false });
  console.log(`Unupdated products: ${unupdatedCount}`);
  if (threshold && unupdatedCount >= threshold) {
    console.log(`Above threshold, aborting.`);
    return;
  }

  try {
    const result = await itemCollection.updateMany({ updated: false }, [
      { $set: { oldprice: "$price" } },
      { $set: { price: 0.0, discount: 0, literprice: 0, alcoholprice: null } },
      { $set: { prices: { $ifNull: ["$prices", []] } } },
      { $set: { prices: { $concatArrays: ["$prices", ["$price"]] } } },
    ]);

    console.log(`Added ${result.modifiedCount} empty prices to unupdated products.`);
  } catch (err) {
    console.error("Error adding unupdated prices:", err);
  }
}

const session = axios.create();

async function main() {
  await itemCollection.updateMany({}, { $set: { updated: false } });
  const alreadyUpdated = await itemCollection
    .find({ updated: true })
    .map((item) => item.index)
    .toArray();

  const startPage = 0;
  await getProducts(startPage, alreadyUpdated);

  // [!] ONLY RUN THIS AFTER ALL PRICES HAVE BEEN UPDATED [!]
  await syncUnupdatedProducts(null);
}

await main();

client.close();
